code/datums/map_generators/_map_generator.dm

```diff
@@ -0,0 +1,30 @@
+/// ======================================
+/// Ruin generator process holder
+/// ======================================
+/datum/map_generator
+	var/completed = FALSE
+	var/ticks = 0
+	var/list/datum/callback/completion_callbacks = list()
+	var/list/callback_args
+
+/// Begin generating
+/datum/map_generator/proc/generate(...)
+	SSmap_generator.executing_generators += src
+	callback_args = args.Copy(1)
+
+/datum/map_generator/proc/on_completion(datum/callback/completion_callback)
+	completion_callbacks += completion_callback
+
+/// Execute a current run.
+/// Returns TRUE if finished
```

Comment:
it returns constant true not true if finished

Reply:
Yea, the default proc which isn't overriden finishes on the first tick since it does nothing, so it always returns true. When overriden, this may take multiple ticks to complete so the comment tells you what should be returned.